lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp

```diff
@@ -1052,43 +1052,170 @@ struct ElementwiseInlineAsmOpConversion
-                                   Type elemTy, MultipleOperandsRange operands,
-                                   Location loc) const {
-    int numPackedElements = op.getPackedElement();
-    if (operands.size() % numPackedElements != 0)
+  SmallVector<SmallVector<Value>>
+  createDestOps(ElementwiseInlineAsmOp op, OpAdaptor adaptor,
+                ConversionPatternRewriter &rewriter,
+                MultipleOperandsRange operands, Location loc) const {
+    auto ctx = op->getContext();
+
+    if (operands.size() % op.getPackedElement() != 0)
       llvm::report_fatal_error("Inline asm op has more packed elements than "
                                "number of elements per thread.");
+
+    // Pack elems smaller than 32 bits into 32-bit registers.
     SmallVector<Value> packedOperands =
         packOperands(op, operands, rewriter, loc);
-    Type dstType =
-        getTypeConverter()->convertType(getElementType(op.getResult()));
-    Type retType = dstType;
-    if (numPackedElements > 1)
-      retType = vec_ty(retType, numPackedElements);
-    Value result = rewriter
-                       .create<LLVM::InlineAsmOp>(
-                           loc, retType,
-                           packedOperands,      // operands
-                           op.getAsmString(),   // asm_string
-                           op.getConstraints(), // constraints
-                           !op.getPure(),       // has_side_effects
-                           false,               // is_align_stack
-                           LLVM::AsmDialectAttr::get(
-                               rewriter.getContext(),
-                               LLVM::AsmDialect::AD_ATT), // asm_dialect
-                           ArrayAttr()                    // operand_attrs
-                           )
-                       ->getResult(0);
-    SmallVector<Value> results;
-    if (numPackedElements > 1) {
-      for (int i = 0; i < numPackedElements; i++)
-        results.push_back(extract_element(result, i32_val(i)));
-    } else {
-      results = {result};
+
+    // Types returned by the LLVM asm op.  If there's more than one, they'll be
+    // wrapped in a struct.
+    SmallVector<Type> asmRetTypes;
+    for (auto result : op.getResult()) {
+      auto ty = getTypeConverter()->convertType(getElementType(result));
+
+      // Pack return elements into 32-bits.
+      unsigned bitWidth = ty.isIntOrFloat() ? ty.getIntOrFloatBitWidth() : 64;
+      unsigned numElemsPerReg =
+          std::min(bitWidth < 32 ? 32 / bitWidth : 1, op.getPackedElement());
+      assert(op.getPackedElement() % numElemsPerReg == 0);
+      if (numElemsPerReg > 1) {
+        ty = vec_ty(ty, numElemsPerReg);
+      }
+      for (unsigned i = 0; i < op.getPackedElement() / numElemsPerReg; i++) {
+        asmRetTypes.push_back(ty);
+      }
     }
-    return results;
+    Type asmRetType =
+        asmRetTypes.size() > 1 ? struct_ty(asmRetTypes) : asmRetTypes[0];
```

Comment:
Could we also support no results as well?

That's a case Adam needed recently and it would be a good occasion to support it as it should generalize. It can also be a follow PR if you think it doesn't make sense to support it in this one.

Reply:
We should definitely do no outputs, but yeah, let me do that as a follow-on.